From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gerson Fernando Budke <gerson.budke@ossystems.com.br>
Subject: [PATCH] [GH#414] STM32_ADC_Common: Add ADC_CALFACT selector

The STM32_ADC_Common define the 0xC4 address for ADC_CALFACT register.
This is only valid for the most advanced version of adc peripheral
which is not compatible with the low cost devices like C0/G0/L0. This
add a selector at class instanciation to allow the correct definition
of the address.

Signed-off-by: Gerson Fernando Budke <gerson.budke@ossystems.com.br>

Ported to Renode v1.17.0 STM32_ADC_Common API.

diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32F0_ADC.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32F0_ADC.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32F0_ADC.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32F0_ADC.cs
@@ @@ public STM32F0_ADC(IMachine machine, double referenceVoltage, uint externalEvent
                 // Base class configuration
                 watchdogCount: 1,
                 hasCalibration: false,
+                hasHighCalAddress: false,
                 channelCount: 19,
                 hasPrescaler: false,
                 hasVbatPin: true,
diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32F3_ADC.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32F3_ADC.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32F3_ADC.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32F3_ADC.cs
@@ @@ public STM32F3_ADC(IMachine machine, double referenceVoltage, uint externalEvent
                 // Base class configuration
                 watchdogCount: 3,
                 hasCalibration: true,
+                hasHighCalAddress: false,
                 channelCount: 19,
                 hasPrescaler: true,
                 hasVbatPin: false,
diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32G0_ADC.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32G0_ADC.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32G0_ADC.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32G0_ADC.cs
@@ @@ public STM32G0_ADC(IMachine machine, double referenceVoltage, uint externalEvent
                 // Base class configuration
                 watchdogCount: 3,
                 hasCalibration: true,
+                hasHighCalAddress: false,
                 channelCount: 19,
                 hasPrescaler: true,
                 hasVbatPin: true,
diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32H7_ADC.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32H7_ADC.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32H7_ADC.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32H7_ADC.cs
@@ @@ public STM32H7_ADC(IMachine machine, double referenceVoltage, uint externalEvent
                 // Base class configuration
                 watchdogCount: 3,
                 hasCalibration: true,
+                hasHighCalAddress: true,
                 channelCount: 19,
                 hasPrescaler: true,
                 hasVbatPin: true,
diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32L5_ADC.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32L5_ADC.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32L5_ADC.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32L5_ADC.cs
@@ @@ public STM32L5_ADC(IMachine machine, double referenceVoltage, uint externalEvent
                 // Base class configuration
                 watchdogCount: 3,
                 hasCalibration: false,
+                hasHighCalAddress: false,
                 channelCount: 19,
                 hasPrescaler: true,
                 hasVbatPin: false,
diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32WBA_ADC.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32WBA_ADC.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32WBA_ADC.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32WBA_ADC.cs
@@ @@ public STM32WBA_ADC(IMachine machine, double referenceVoltage, uint externalEven
                 // Base class configuration
                 watchdogCount: 3,
                 hasCalibration: true,
+                hasHighCalAddress: true,
                 channelCount: 14,
                 hasPrescaler: true,
                 hasVbatPin: false,
diff --git a/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs b/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs
--- a/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs
+++ b/src/Emulator/Peripherals/Peripherals/Analog/STM32_ADC_Common.cs
@@ @@
 //
 // Copyright (c) 2010-2026 Antmicro
+// Copyright (c) 2023 OS Systems
 //
 // This file is licensed under the MIT License.
 // Full license text is available in 'licenses/MIT.txt'.
@@ @@ namespace Antmicro.Renode.Peripherals.Analog
     //     watchdogCount ------ Specifies the number of analog watchdogs inside the peripheral between 1 and 3.
     //    *hasCalibration ----- Specifies whether the calibration factor and voltage regulator are available to the software.
     //                          ADCs without this feature will still have the ADCAL flag available to trigger the calibration procedure,
-    //                          but not the CALFACT register.
+    //                          but not the CALFACT register nor the ADVREGEN field.
+    //     hasHighCalAddress -- Specifies that ADC_CALFACT is at low address or high address.
     //     channelCount ------- Specifies the amount of available channels.
     //                          Includes both internal sources (like the temperature sensor) as well as external.
     //    *hasPrescaler ------- Specifies whether the ADC contains a prescaler for the external clock input.
@@ @@ namespace Antmicro.Renode.Peripherals.Analog
     public abstract class STM32_ADC_Common : IKnownSize, IProvidesRegisterCollection<DoubleWordRegisterCollection>, IDoubleWordPeripheral, IWordPeripheral, IADC
     {
         public STM32_ADC_Common(IMachine machine, double referenceVoltage, uint externalEventFrequency, int dmaChannel, IDMA dmaPeripheral,
-            int watchdogCount, bool hasCalibration, int channelCount, bool hasPrescaler,
+            int watchdogCount, bool hasCalibration, bool hasHighCalAddress, int channelCount, bool hasPrescaler,
             bool hasVbatPin, bool hasChannelSequence, bool hasPowerRegister, bool hasChannelSelect,
             bool hasOffset, bool hasDifferentialMode, SamplingTime samplingTime, bool dualMode, bool hasLinearityCalibration, bool hasChannelInjection, bool hasSeparateThresholdRegisters, ResolutionRange resolutionRange, bool hasChannelPreselection, bool hasScanDirection)
         {
@@ @@ public STM32_ADC_Common(IMachine machine, double referenceVoltage, uint external
             }
 
             registers = new DoubleWordRegisterCollection(this, BuildRegistersMap(hasCalibration,
+                                                                                 hasHighCalAddress,
                                                                                  hasPrescaler,
                                                                                  hasVbatPin,
                                                                                  hasChannelSequence,
@@ @@ private uint MillivoltsToSample(double sampleInMillivolts, Resolution sampleReso
             return referencedValue;
         }
 
-        private Dictionary<long, DoubleWordRegister> BuildRegistersMap(bool hasCalibration, bool hasPrescaler, bool hasVbatPin, bool hasChannelSequence, bool hasPowerRegister, bool hasOffset, bool hasDifferentialMode, SamplingTime samplingTime, bool dualMode, bool hasLinearityCalibration, bool hasChannelInjection, bool hasSeparateThresholdRegisters, bool hasScanDirection)
+        private Dictionary<long, DoubleWordRegister> BuildRegistersMap(bool hasCalibration, bool hasHighCalAddress, bool hasPrescaler, bool hasVbatPin, bool hasChannelSequence, bool hasPowerRegister, bool hasOffset, bool hasDifferentialMode, SamplingTime samplingTime, bool dualMode, bool hasLinearityCalibration, bool hasChannelInjection, bool hasSeparateThresholdRegisters, bool hasScanDirection)
         {
             var isrRegister = new DoubleWordRegister(this)
                 .WithFlag(0, out adcReadyFlag, FieldMode.Read | FieldMode.WriteOneToClear, name: "ADRDY")
@@ @@ private Dictionary<long, DoubleWordRegister> BuildRegistersMap(bool hasCalibrati
 
             if(hasCalibration)
             {
-                registers.Add((long)Registers.CalibrationFactor, new DoubleWordRegister(this)
+                var calibrationFactorRegister = hasHighCalAddress ? Registers.CalibrationFactorHigh : Registers.CalibrationFactorLow;
+                registers.Add((long)calibrationFactorRegister, new DoubleWordRegister(this)
                     .WithValueField(0, 7, name: "CALFACT")
                     .WithReservedBits(7, 25));
             }
@@ @@ private enum Registers
             DifferentialMode       = 0xB0, // ADC_DIFSEL
             WatchdogLowThreshold2  = 0xB0, // ADC_LTR2
             WatchdogHighThreshold2 = 0xB4, // ADC_HTR2
+            CalibrationFactorLow   = 0xB4, // ADC_CALFACT (C0/F3/G0/L0)
             WatchdogLowThreshold3  = 0xB8, // ADC_LTR3
             WatchdogHighThreshold3 = 0xBC, // ADC_HTR3
             // Gap intended
             DifferentialMode2      = 0xC0, // ADC_DIFSEL
-            CalibrationFactor      = 0xC4, // ADC_CALFACT
+            CalibrationFactorHigh  = 0xC4, // ADC_CALFACT (H7/WBA)
             CalibrationFactor2     = 0xC8, // ADC_CALFACT2
             // Gap intended
             CommonConfiguration    = 0x308, // ADC_CCR
